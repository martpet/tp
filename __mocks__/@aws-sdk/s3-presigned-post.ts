export const createPresignedPost = vi.fn().mockResolvedValue('dummyPresignedPost');
