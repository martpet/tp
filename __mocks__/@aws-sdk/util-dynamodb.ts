export const unmarshall = vi
  .fn()
  .mockName('unmarshall')
  .mockReturnValue('dummyUnmarshalledItems');
