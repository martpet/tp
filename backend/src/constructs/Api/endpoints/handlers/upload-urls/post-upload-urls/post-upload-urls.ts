import {
  APIGatewayProxyHandlerV2,
  createPresignedPost,
  EnvVars,
  errorResponse,
  getIdTokenPayload,
  maxPhotoUploadSize,
  PostUploadUrlsRequest,
  PostUploadUrlsResponse,
  RouteHeaders,
  S3Client,
  StatusCodes,
} from 'lambda-layer';

import { findExistingFingerprints } from '~/constructs/Api/utils';

const s3Client = new S3Client({});

export const handler: APIGatewayProxyHandlerV2<PostUploadUrlsResponse> = async (
  event
) => {
  const { authorization } = event.headers as RouteHeaders<'/settings'>;
  const { photoBucket } = process.env as EnvVars<'/upload-urls', 'POST'>;

  if (!authorization) {
    return errorResponse('Vf5Ph6qN1S');
  }

  if (!photoBucket) {
    return errorResponse('1rQkj3kpp4');
  }

  if (!event.body) {
    return errorResponse('G-luuHqI0s', { statusCode: StatusCodes.BAD_REQUEST });
  }

  let requestData;

  try {
    requestData = JSON.parse(event.body) as PostUploadUrlsRequest;
  } catch (error) {
    return errorResponse('9210145fdf', { statusCode: StatusCodes.BAD_REQUEST, error });
  }

  const existingFingerprintsInDb = await findExistingFingerprints(
    requestData.map(({ fingerprint }) => fingerprint)
  );

  const uniqueItems = requestData.filter(
    ({ fingerprint }) => !existingFingerprintsInDb.includes(fingerprint)
  );

  const { sub } = await getIdTokenPayload(authorization);

  const uploadUrlsEntries = await Promise.all(
    uniqueItems.map(async ({ id, fingerprint, digest }) => {
      const presignedPost = await createPresignedPost(s3Client, {
        Bucket: photoBucket,
        Key: `${sub}/${fingerprint}`,
        Expires: 600,
        Fields: {
          'x-amz-checksum-algorithm': 'SHA256',
          'x-amz-checksum-sha256': digest,
        },
        Conditions: [['content-length-range', 0, maxPhotoUploadSize]],
      });
      return [id, presignedPost];
    })
  );

  return {
    presignedPosts: Object.fromEntries(uploadUrlsEntries),
    existingFingerprintsInDb,
  };
};
