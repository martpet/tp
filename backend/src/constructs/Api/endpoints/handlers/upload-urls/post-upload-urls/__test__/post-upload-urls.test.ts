import { createPresignedPost } from '@aws-sdk/s3-presigned-post';

import {
  findExistingFingerprints,
  itCalls,
  itGetsIdToken,
  itHasEnvVars,
  itHasJsonBody,
  itResolves,
} from '~/constructs/Api/utils';

import { handler } from '../post-upload-urls';

vi.mock('@aws-sdk/s3-presigned-post');
vi.mock('@aws-sdk/client-s3');
vi.mock('~/constructs/Api/utils/errorResponse');
vi.mock('~/constructs/Api/utils/getIdTokenPayload');
vi.mock('~/constructs/Api/utils/XfindExistingFingerprints');

process.env.photoBucket = 'dummyPhotoBucket';

const args = [
  {
    headers: { authorization: 'dummyAuthorizationHeader' },
    body: JSON.stringify([
      {
        id: 'dummyId1',
        fingerprint: 'dummyFingerprint1',
        hash: 'dummyHash1',
      },
      {
        id: 'dummyId2',
        fingerprint: 'dummyFingerprint2',
        hash: 'dummyHash2',
      },
    ]),
  },
] as unknown as Parameters<typeof handler>;

describe('post-upload-urls', () => {
  itHasJsonBody(handler, args);
  itHasEnvVars(['photoBucket'], handler, args);
  itGetsIdToken(handler, args);
  itResolves(handler, args);
  itCalls(createPresignedPost, handler, args);
  itCalls(findExistingFingerprints, handler, args);
});
