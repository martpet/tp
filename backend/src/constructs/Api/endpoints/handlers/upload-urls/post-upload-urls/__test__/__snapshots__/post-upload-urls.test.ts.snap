// Vitest Snapshot v1

exports[`post-upload-urls > calls "createPresignedPost" with correct args 1`] = `
[
  [
    DummyS3Client {},
    {
      "Bucket": "dummyPhotoBucket",
      "Conditions": [
        [
          "content-length-range",
          0,
          31457280,
        ],
      ],
      "Expires": 10,
      "Fields": {
        "x-amz-checksum-algorithm": "SHA256",
        "x-amz-checksum-sha256": undefined,
      },
      "Key": "dummySub/dummyFingerprint2.jpg",
    },
  ],
]
`;

exports[`post-upload-urls > calls "findExistingFingerprints" with correct args 1`] = `
[
  [
    [
      "dummyFingerprint1",
      "dummyFingerprint2",
    ],
  ],
]
`;

exports[`post-upload-urls > calls "getIdTokenPayload" with correct args 1`] = `
[
  [
    "dummyAuthorizationHeader",
  ],
]
`;

exports[`post-upload-urls > resolves with a correct value 1`] = `
{
  "existingFingerprintsInDb": [
    "dummyFingerprint1",
  ],
  "presignedPosts": {
    "dummyId2": "dummyPresignedPost",
  },
}
`;

exports[`post-upload-urls > when "authorization" header is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "Vf5Ph6qN1S",
  ],
]
`;

exports[`post-upload-urls > when "authorization" header is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-upload-urls > when "event.body" is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "G-luuHqI0s",
    {
      "statusCode": 400,
    },
  ],
]
`;

exports[`post-upload-urls > when "event.body" is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-upload-urls > when "event.body" is not JSON > calls "errorResponse" with correct args 1`] = `
[
  [
    "9210145fdf",
    {
      "error": [SyntaxError: Unexpected token o in JSON at position 1],
      "statusCode": 400,
    },
  ],
]
`;

exports[`post-upload-urls > when "event.body" is not JSON > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-upload-urls > when "photoBucket" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "1rQkj3kpp4",
  ],
]
`;

exports[`post-upload-urls > when "photoBucket" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-upload-urls > when \`findExistingFingerprints\` throws > calls "errorResponse" with correct args 1`] = `
[
  [
    "9fb96f4182",
    {
      "error": [Error: dummyFindExistingItemsError],
    },
  ],
]
`;

exports[`post-upload-urls > when \`findExistingFingerprints\` throws > resolves with a correct value 1`] = `"dummyErrorResponse"`;
